homeassistant/components/shelly/update.py

```diff
@@ -0,0 +1,198 @@
+        entity_registry_enabled_default=False,
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up update entities for Shelly component."""
+    if get_device_entry_gen(config_entry) == 2:
+        return async_setup_entry_rpc(
+            hass, config_entry, async_add_entities, RPC_UPDATES, RpcUpdateEntity
+        )
+
+    if not config_entry.data[CONF_SLEEP_PERIOD]:
+        async_setup_entry_rest(
+            hass,
+            config_entry,
+            async_add_entities,
+            REST_UPDATES,
+            RestUpdateEntity,
+        )
+
+
+class RestUpdateEntity(ShellyRestAttributeEntity, UpdateEntity):
+    """Represent a REST update entity."""
+
+    _attr_supported_features = UpdateEntityFeature.INSTALL
+    entity_description: RestUpdateDescription
+
+    @property
+    def installed_version(self) -> str | None:
+        """Version currently in use."""
+        version = self.wrapper.device.status["update"]["old_version"]
+        if version is None:
+            return None
+
+        return str(version)
+
+    @property
+    def latest_version(self) -> str | None:
+        """Latest version available for install."""
+        new_version = self.entity_description.latest_version(
+            self.wrapper.device.status,
+        )
+        if new_version is not None:
+            return str(new_version)
+
+        return self.installed_version
+
+    async def async_install(
+        self, version: str | None, backup: bool, **kwargs: Any
+    ) -> None:
+        """Install the latest firmware version."""
+        config_entry = self.wrapper.entry
+        block_wrapper = self.hass.data[DOMAIN][DATA_CONFIG_ENTRY][
+            config_entry.entry_id
+        ].get(BLOCK)
+        await self.entity_description.install(block_wrapper)
+
+
+class RpcUpdateEntity(ShellyRpcAttributeEntity, UpdateEntity):
+    """Represent a RPC update entity."""
+
+    _attr_supported_features = UpdateEntityFeature.INSTALL
+    entity_description: RpcUpdateDescription
+
+    @property
+    def installed_version(self) -> str | None:
+        """Version currently in use."""
+        if self.wrapper.device.shelly is None:
+            return None
+
+        return str(self.wrapper.device.shelly["ver"])
+
+    @property
+    def latest_version(self) -> str | None:
+        """Latest version available for install."""
+        new_version = self.entity_description.latest_version(
+            self.wrapper.device.status[self.key][self.entity_description.sub_key],
+        )
+        if new_version is not None:
+            return str(new_version)
+
+        return self.installed_version
+
+    async def async_install(
+        self, version: str | None, backup: bool, **kwargs: Any
+    ) -> None:
+        """Install the latest firmware version."""
+        await self.entity_description.install(self.wrapper)
```

Comment:
Can we track the installation progress based on NotifyEvent?
```json
{"src":"shellypro1pm-xxxxxx80e0c0","dst":"shellypro1pm-xxxxxx80e0c0/events","method":"NotifyEvent","params":{"ts":1662043461.52,"events":[{"component":"sys", "event":"ota_progress", "msg":"Waiting for data", "progress_percent":25, "ts":1662043461.52}]}}
```

Reply:
Probably, but I am not so familiar with the shelly code, so I would like to leave that to a separate PR.
Maybe someone from the shelly codeoweners is willing to implement that?

Reply:
I implemented the basic True/False logic, however I can't test it because I updated all my shelly devices in testing this code.
So I don't have any updates left.
I know the rest works.

Reply:
For Gen1 devices you can downgrade using http://archive.shelly-tools.de/
I am using it for testing, just make sure to take your correct device model and it is also recommended to downgrade only 1 version back to eliminate change in device settings.